### DISCUSSION TITLE
Beispiel-Code möglich? Wie kann ich mich authentifizieren?

### DISCUSSION THREAD

**Discussion by tcsum** [Digitaler-Impfnachweis/certification-apis#17] [Q&A] [answered]
Hallo zusammen 

Ich bin jetzt an der API und habe mir ein p12 Zertfikat generiert. Das scheint auch zu klappen.

Jetzt hängt es aber hier:

     RestClient::NotFound:   404 Not Found.

URL:

https://api.certify.demo.ubirch.com/api/certify/v2/issue

Ich schicke noch keinen JWT mit. Was ich mich frage: warum gibt es einen 404? Das müsste doch nen Not Authorized geben, oder?

Und: wie bekomme ich den JWT? Die Dokumentation ist hier etwas dünn.

--> Gibt es irgendeine Chance, dass ihr uns code für 'nen client zur Verfügung stellt? Oder cURL Aufrufe, etc. Irgendwas, was auch für einen Senior Entwickler zu verstehen ist?

**Comment by oliver-steinbrecher** [comment]
Ich habe es gerade mit curl getestet . Ich erhalte hier einen Zertifikatsfehler, weil ich kein client-cert übergeben habe. Der Endpunkt ist aber grundsätzlich erreichbar:

```
curl -v https://api.certify.demo.ubirch.com/api/certify/v2/issue
*   Trying 80.158.62.207...
* TCP_NODELAY set
* Connected to api.certify.demo.ubirch.com (80.158.62.207) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Request CERT (13):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Certificate (11):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS alert, bad certificate (554):
* error:1401E412:SSL routines:CONNECT_CR_FINISHED:sslv3 alert bad certificate
* Closing connection 0
curl: (35) error:1401E412:SSL routines:CONNECT_CR_FINISHED:sslv3 alert bad certificate
```

**Comment by thinkberg** [comment, marked as answer]
Nach direkter Kommunikation: Bitte nochmal code testen. Die Zertifikate funktionieren nach meiner Ueberpruefung.

**Comment by tcsum** [comment]
Es funktioniert jetzt.

Die PDF-Rückgabe ist aber noch nicht implementiert, richtig?

(Für uns okay - wir generieren schon ein brauchbares PDF. Wir waren nur mal neugierig)

**Reply by thinkberg** [reply]
Korrekt. Ist in Arbeit.

**Comment by tcsum** [comment]
Ich habe jetzt wieder Probleme. Jeweils mit dem gleichen cert+pass:

Bei uns auf Stage:

=> "#<RestClient::BadRequest: 400 Bad Request>"
irb(main):113:0> err.http_body
=> "Sorry, there is something invalid in your request. Error Validating with Schema"

```
 {:nam=>{:fn=>"Testzwei", :gn=>"Impfpass"}, :dob=>"1876-05-22", :v=>[{:id=>"IZ28215B", :tg=>"840539006", :vp=>"1119305005", :mp=>"EU/1/21/1529", :ma=>"ORG-100001699", :dn=>1, :sd=>1, :dt=>"2021-05-21"}]}
```

Auf development:

```
=> "#<RestClient::Forbidden: 403 Forbidden>"
irb(main):020:0> err.http_body
=> "Sorry, the credential is invalid."
```

Komisch, genau mit dem code hat es vor ein paar Tagen noch funktioniert. Irgendeine Idee?

**Reply by thinkberg** [reply]
Beim Credential koennte es sein, dass der Client nicht die Zertifikatskette mitsendet, das wird inzwischen geprueft. Bitte dort mal nachsehen. Wegen Schema muss ich nachsehen.

**Reply by thinkberg** [reply]
Das Geburtsdatum ist falsch: `:dob=>"1876-05-22"` 
Laut spec muss das Geb. Datum nach 1920 sein.

**Reply by thinkberg** [reply]
Das client cert habe ich eben noch einmal geprueft. Das funktioniert auf jeden Fall, bitte checken, dass das komplette Zertifikat gesendet wird.

https://github.com/Digitaler-Impfnachweis/certification-apis/blob/master/dgc-schema.json#L32

**Comment by tcsum** [comment]
> Das Geburtsdatum ist falsch: :dob=>"1876-05-22"

check!

**Reply by tcsum** [reply]
Bei Ruby, das war hilfreich: https://bugs.ruby-lang.org/issues/9758

Das auch: https://medium.com/in-the-weeds/net-http-and-x509-client-certificate-chains-and-monkey-patches-oh-my-ea5258dcb697 

Allerdings geht es noch nicht. Wir suchen noch, wie man das in Ruby reinkriegt. 

Falls ihr hier 'ne Idee habt: gerne schreiben.

Ansonsten: der cURL call geht auch nicht mehr, könnt ihr den noch mal schicken?

**Reply by tcsum** [reply]
Wenn man mal ins p12 guckt:

`openssl pkcs12 -in ./stage_ubirch_Group_Access_sumcumo-hh-vac.p12 -nodes -passin pass:"fef_______2" `

--> da ist doch nur ein Zertifikat drin! Wie sollen wir dann die chain schicken?
